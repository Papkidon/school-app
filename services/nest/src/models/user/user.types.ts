import { z } from 'zod';

export const userCreateSchema = z
  .object({
    data: z.object({
      username: z.string().min(2).max(20),
      password: z.string().min(8).max(20),
    }),
  })
  .strict();

export const userUpdateSchema = userCreateSchema.deepPartial();

export type userType = {
  data: {
    username: string;
    password: string;
  };
};
