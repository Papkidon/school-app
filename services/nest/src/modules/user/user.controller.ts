import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import {
  userCreateSchema,
  userUpdateSchema,
} from '../../models/user/user.types';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { User } from '@prisma/client';
import { userType } from '../../models/user/user.types';
import { DeepPartial, idSchema } from '../../models/utility.types';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import { UserService } from './user.service';

@Controller('api/v1/user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get('')
  async findAll(): Promise<User[]> {
    try {
      const result = await this.userService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No user data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(userCreateSchema)) createUserDto: userType,
  ): Promise<User> {
    try {
      const userExists = await this.userService.findByUsername(
        createUserDto.data.username,
      );

      if (userExists) {
        throw new HttpException(
          'User with given username already exists',
          HttpStatus.CONFLICT,
        );
      }

      return await this.userService.create(createUserDto);
    } catch (error) {
      throw error;
    }
  }

  @Post('random')
  async createRandomUser() {
    try {
      return await this.userService.createRandomUser();
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<User> {
    try {
      const result = await this.userService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(userUpdateSchema))
    body: DeepPartial<userType>,
  ): Promise<User> {
    try {
      const userExists = await this.userService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'User with given username does not exist',
          HttpStatus.NOT_FOUND,
        );
      }
      return await this.userService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<User> {
    try {
      const userExists = await this.userService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'User with given username does not exist',
          HttpStatus.NOT_FOUND,
        );
      }
      return await this.userService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
