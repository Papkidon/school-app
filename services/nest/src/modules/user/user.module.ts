import { Module } from '@nestjs/common';
import { UserController } from './user.controller';
import { UserService } from './user.service';
import { UserRepository } from './user.repository';
import { LoggerModule } from 'nestjs-pino';
import { PrismaModule, PrismaService } from 'nestjs-prisma';

@Module({
  imports: [
    LoggerModule.forRoot({
      pinoHttp: {
        transport: {
          target: 'pino-pretty',
          options: {
            singleLine: true,
          },
        },
      },
    }),
    PrismaModule.forRoot(),
  ],
  controllers: [UserController],
  providers: [UserService, UserRepository, PrismaService],
})
export class UserModule {}
