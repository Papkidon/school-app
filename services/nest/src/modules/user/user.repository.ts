import { Injectable } from '@nestjs/common';
import { User } from '@prisma/client';
import { PrismaService } from 'nestjs-prisma';
import { userType } from '../../models/user/user.types';
import { DeepPartial } from '../../models/utility.types';

@Injectable()
export class UserRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<User[]> {
    return await this.prisma.user.findMany();
  }

  async findById(id: string): Promise<User> {
    return await this.prisma.user.findUnique({ where: { id } });
  }

  async findByUsername(username: string): Promise<User> {
    return await this.prisma.user.findUnique({ where: { username } });
  }

  async create(body: any): Promise<User> {
    return await this.prisma.user.create(body);
  }

  async createRandomUser(username: string, password: string): Promise<User> {
    return await this.prisma.user.create({
      data: {
        username,
        password,
      },
    });
  }

  async delete(id: string): Promise<User> {
    return await this.prisma.user.delete({ where: { id } });
  }

  async update(id: string, body: DeepPartial<userType>): Promise<User> {
    return await this.prisma.user.update({
      where: { id },
      data: { ...body.data },
    });
  }
}
