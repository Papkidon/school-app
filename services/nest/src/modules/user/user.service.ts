import { Injectable } from '@nestjs/common';
import { UserRepository } from './user.repository';
import { userType } from '../../models/user/user.types';
import { DeepPartial } from '../../models/utility.types';
import { User } from '@prisma/client';
import BuildPerson from '../../utils/BuildPerson/BuildPerson';

@Injectable()
export class UserService {
  constructor(private userRepository: UserRepository) {}
  async findAll(): Promise<User[]> {
    return await this.userRepository.findAll();
  }

  async findById(id: string): Promise<User> {
    return await this.userRepository.findById(id);
  }

  async create(body: any): Promise<User> {
    return await this.userRepository.create(body);
  }

  async createRandomUser(): Promise<User> {
    const randomUser = new BuildPerson().randomize().build();
    return await this.userRepository.createRandomUser(...randomUser);
  }

  async delete(id: string): Promise<User> {
    return await this.userRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<userType>): Promise<User> {
    return await this.userRepository.update(id, body);
  }
}
