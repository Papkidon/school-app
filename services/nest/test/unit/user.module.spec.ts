import { UserController } from '../../src/modules/user/user.controller';
import { UserService } from '../../src/modules/user/user.service';
import { UserRepository } from '../../src/modules/user/user.repository';
import { PrismaService } from 'nestjs-prisma';
import { User } from '@prisma/client';
import {
  dummyUserId,
  dummyUserUpdate,
  dummyUserCreate,
} from './utils/dummy.data';

jest.mock('uuid', () => ({ v4: () => 'cdd77c0f-63f6-49a8-8023-dd05f8ec5dcf' }));
jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

describe('UserController', () => {
  let userController: UserController;
  let userService: UserService;
  let userRepository: UserRepository;
  let prismaService: PrismaService;

  beforeEach(() => {
    prismaService = new PrismaService();
    userRepository = new UserRepository(prismaService);
    userService = new UserService(userRepository);
    userController = new UserController(userService);
  });

  afterAll(() => {
    jest.useRealTimers();
    jest.resetAllMocks();
  });

  describe('findAll', () => {
    it('should return an array of all users', async () => {
      const result = ['user1', 'user2'] as unknown as User[];
      jest
        .spyOn(userRepository, 'findAll')
        .mockImplementation(async () => result);

      expect(await userController.findAll()).toMatchSnapshot();
    });
  });

  describe('findUnique', () => {
    it('should return an user with given id', async () => {
      const result = 'dummyFoundByIdUser' as unknown as User;
      jest
        .spyOn(userRepository, 'findById')
        .mockImplementation(async () => result);

      expect(await userController.findById(dummyUserId)).toMatchSnapshot();
    });
  });

  describe('create', () => {
    it('should create a new user', async () => {
      const result = 'dummyCreatedUser' as unknown as User;
      jest
        .spyOn(userRepository, 'create')
        .mockImplementation(async () => result);

      expect(await userController.create(dummyUserCreate)).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('should update an user with given id', async () => {
      const result = 'dummyUpdatedUser' as unknown as User;
      jest
        .spyOn(userRepository, 'update')
        .mockImplementation(async () => result);

      expect(
        await userController.update(dummyUserId, dummyUserUpdate),
      ).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('should delete an user with given id', async () => {
      const result = 'dummyDeletedUser' as unknown as User;
      jest
        .spyOn(userRepository, 'delete')
        .mockImplementation(async () => result);

      expect(await userController.delete(dummyUserId)).toMatchSnapshot();
    });
  });
});
