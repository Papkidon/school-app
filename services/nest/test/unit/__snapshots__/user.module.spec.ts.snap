// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController create should create a new user 1`] = `"dummyCreatedUser"`;

exports[`UserController delete should delete an user with given id 1`] = `"dummyDeletedUser"`;

exports[`UserController findAll should return an array of all users 1`] = `
[
  "user1",
  "user2",
]
`;

exports[`UserController findUnique should return an user with given id 1`] = `"dummyFoundByIdUser"`;

exports[`UserController update should update an user with given id 1`] = `"dummyUpdatedUser"`;
